src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunTableManager.cs

```diff
@@ -54,22 +70,96 @@ protected virtual void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
                 lock (_methodsGenerated)
                 {
                     Debug.Assert(!_sortedMethods);
-                    _methodsGenerated.Add(methodNode.Method);
+                    MethodDesc method = methodNode.Method;
+                    EcmaModule module = (EcmaModule)((EcmaMethod)method.GetTypicalMethodDefinition()).Module;
+                    if (!_methodsGenerated.TryGetValue(module, out var perModuleData))
+                    {
+                        perModuleData = new PerModuleMethodsGenerated(module);
+                        _methodsGenerated[module] = perModuleData;
+                    }
+                    if (method.HasInstantiation || method.OwningType.HasInstantiation)
+                    {
+                        perModuleData.GenericMethodsGenerated.Add(methodNode);
+                    }
+                    else
+                    {
+                        perModuleData.MethodsGenerated.Add(methodNode);
+                    }
                 }
             }
         }
-        public IEnumerable<MethodDesc> GetCompiledMethods()
+
+        public IEnumerable<IMethodNode> GetCompiledMethods(EcmaModule module, bool? genericInstantiations)
         {
             lock (_methodsGenerated)
```

Comment:
Would there be an advantage to introduce double-checked locking here to avoid waiting on the lock if methods are already sorted?

Reply:
No, this method isn't called often enough for it to matter.